### ISSUE TITLE
Unexpected InvalidReturnStatement with methods of generic classes calling out to methods returning static

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#2439]
_From the discussion in #2433_ 

Below, for any descendant of `Base`, `factory()` is expected to return an instance of the same class it was called on (and is flagged by Psalm if it doesn't). So where a type parameter is constrained to be `T of Base` we can assume `T::factory()` returns `T`. Thus `C::f()` should not be flagged with `InvalidReturnStatement`.

https://psalm.dev/r/1a056c3be2
```php
<?php
class Base {
  /** @return static */
  public static function factory(): self {
    return new self();
  }
}

/** 
 * @template T of Base
 * @param class-string<T> $t
 * @return T 
 */
function f(string $t) {
  return $t::factory(); // in functions it's fine
}

/** @template T of Base */
class C {
  /** @var class-string<T> */
  private string $t;

  /** @param class-string<T> $t */
  public function __construct($t) { $this->t = $t; }
  
  /** @return T */
  public function f(): Base {
    return ($this->t)::factory(); // and it should be allowed in methods just like it is for the function above
  }
}
```
**Expected**: no issues
**Actual**:
```
Psalm output (using commit 2b6831b):

ERROR: InvalidReturnStatement - 27:12 - The type 'T as Base' does not match the declared return type 'T:C as Base' for C::f

ERROR: InvalidReturnType - 25:15 - The declared return type 'T' for C::f is incorrect, got 'T'
```

**Comment by ste93cry** [comment]
I have the same problem with [this](
https://psalm.dev/r/7f375e5384) example, although `static` in my example does not appear and so it's not the cause. Let me know if I need to open a separate issue or if I'm missing something and there is no bug at all in my case

**Comment by muglug** [comment]
While this would hold for a regular, non-generic class, but when the value passed to the template is something like `Collection<Foo>`, the class name is not enough to reconstruct all its data.

**Comment by ste93cry** [comment]
Since we posted at the same exact time, can you please confirm your reply was referring to my example too?

**Comment by muglug** [comment]
@ste93cry not a bug, in your case - I'd advise something like https://psalm.dev/r/bf583821bc

You're not the first to ask for a similar construction - see https://github.com/vimeo/psalm/issues/1969 and https://github.com/phpstan/phpstan/issues/2592#issuecomment-552496442.

If you want to ticket that as a new feature request, feel free - you may want to tag `@aidantwoods` and `@TysonAndre` (he may have opinions)

**Comment by ste93cry** [comment]
I didn't think at all that the variable had to be declared with `class-string` as type of the keys of the array, but it's absolutely logical! Sorry for bothering you with a false positive.

**Comment by muglug** [comment]
@ste93cry the difference isn't the `class-string` annotation (though I think that is desired). The difference is in the `/** @var ... */` docbblock above the `@return`.

**Comment by ste93cry** [comment]
You're right. I though that the `class-string` made the real difference, but removing that annotation effectively makes the issue surface again. I also think that these examples are not edge cases or uncommon cases so a proper solution would be indeed cool. Until that happen, thank you for sharing once again how to workaround the issue

**Comment by weirdan** [comment]
> While this would hold for a regular, non-generic class, when the value passed to the template is something like `Collection<Foo>`, the class name is not enough to reconstruct all its data.

@muglug  `Base` is non-generic here, so I don't see how that applies.

**Comment by muglug** [comment]
Well, you could have

```php
/**
 * @template T
 */
class BaseTemplated extends Base {
}
```

and I was thinking there was a way for that information to get lost, but maybe not. Going to reopen
